crates/wasmtime/src/store.rs

```diff
@@ -1855,17 +1868,26 @@ unsafe impl<T> wasmtime_runtime::Store for StoreInner<T> {
     }
 
     fn new_epoch(&mut self) -> Result<u64, anyhow::Error> {
-        return match &self.epoch_deadline_behavior {
-            &EpochDeadline::Trap => {
+        return match &mut self.epoch_deadline_behavior {
+            &mut EpochDeadline::Trap => {
                 let trap = Trap::new_wasm(
                     None,
                     wasmtime_environ::TrapCode::Interrupt,
                     wasmtime_runtime::Backtrace::new(),
                 );
                 Err(anyhow::Error::from(trap))
             }
+            EpochDeadline::Callback(callback) => match callback(&mut self.data) {
+                Ok(delta) => {
+                    // Set a new deadline and return the new epoch deadline so
+                    // the Wasm code doesn't have to reload it.
+                    self.set_epoch_deadline(delta);
+                    Ok(self.get_epoch_deadline())
+                }
+                Err(e) => Err(e),
+            },
             #[cfg(feature = "async")]
-            &EpochDeadline::YieldAndExtendDeadline { delta } => {
+            &mut EpochDeadline::YieldAndExtendDeadline { delta } => {
```

Comment:
For symmetry with above this `&mut` can also be removed although you'll need to do `let delta = *delta` below I think